### PR TITLE
feat: Add encode_max for EthMessage

### LINKED ISSUE #6083: Add EthMessage::encode_max
### Describe the feature

Add a new encode function that ensures the encoded message is smaller than a given size:

https://github.com/paradigmxyz/reth/blob/9d65a4cfbb39d0fc9a90ef9aee451595bcd05ba5/crates/net/eth-wire/src/types/message.rs#L170-L170

https://github.com/paradigmxyz/reth/blob/9d65a4cfbb39d0fc9a90ef9aee451595bcd05ba5/crates/net/eth-wire/src/types/message.rs#L217-L218

This could be implemented as Extension trait for `Vec<Encodable>` 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by supernovahs**
closes #6083

### CODE REVIEW

File: crates/net/eth-wire/src/types/message.rs [hunk 1]
```diff
@@ -214,6 +214,23 @@ impl EthMessage {
     }
 }
 
+trait EncodableExt {
+    fn encode_max(&self, size: usize) -> alloy_rlp::Result<Vec<u8>, String>;
+}
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:219]: we want two functions here,

one that returns an error and one that truncates: put as many encoded items in the buffer as possible

File: crates/net/eth-wire/src/types/message.rs [hunk 2]
```diff
@@ -214,6 +214,23 @@ impl EthMessage {
...
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>, String> {
+        let mut buffer = Vec::new();
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return Err("Size limit exceeded".to_string());
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:227]: I'm okay with returning a `DecodeError` here

**Review by mattsse — CHANGES_REQUESTED** [review]
@DaniPopes is this something we want to upstream to alloy-rlp?

File: crates/net/eth-wire/src/types/message.rs [hunk 3]
```diff
@@ -214,6 +216,39 @@ impl EthMessage {
...
+}
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>, P2PStreamError> {
+        let mut buffer = Vec::new();
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return Err(P2PStreamError::HandshakeError(P2PHandshakeError::DecodeError(
+                    alloy_rlp::Error::Custom("Size limit exceeded"),
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:232]: let's return alloy_rlp::Error directly here

File: crates/net/eth-wire/src/types/message.rs [hunk 4]
```diff
@@ -214,6 +216,39 @@ impl EthMessage {
...
+                )));
+            }
+        }
+        Ok(buffer)
+    }
+
+    fn encode_truncate(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>> {
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:239]: This should be infallible so return Vec directly

File: crates/net/eth-wire/src/types/message.rs [hunk 5]
```diff
@@ -214,6 +216,39 @@ impl EthMessage {
...
+        let mut buffer = Vec::new();
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                buffer.truncate(limit);
+                break;
+            }
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:246]: this is slightly more complex

the logic should be:
encode all items into buffer if putting an item into buffer leads to: limit, split off the last item and return the Vec
so we need to keep track of the len of the buffer before inserting a new item, then split it off there if it exceeds

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by emhane** [comment]
lgtm. this is half of what's needed to close https://github.com/paradigmxyz/reth/issues/6083. the next half is plugging it into https://github.com/paradigmxyz/reth/blob/720815a83465a4c79c0066e466ad53d005cc32c1/crates/net/network/src/eth_requests.rs#L181-L181, then we're done!

**Comment by emhane** [comment]
> The goal is to remove the below check and instead use encode_max right?
> 
> https://github.com/paradigmxyz/reth/blob/720815a83465a4c79c0066e466ad53d005cc32c1/crates/net/network/src/eth_requests.rs#L181-L185

yeah that's right, but it would be good to do it before we fetch the block from store

**Comment by emhane** [comment]
lgtm, make sure to update docs too which is using that old code https://github.com/paradigmxyz/reth/blob/main/docs/crates/network.md

File: crates/net/eth-wire/src/types/message.rs [hunk 6]
```diff
@@ -214,6 +213,44 @@ impl EthMessage {
...
+/// to encode data with size constraints. It is designed to work with collections of encodable
+/// items.
+pub trait EncodableExt {
+    /// Encodes the data, ensuring that the encoded message does not exceed the specified size
+    /// limit. If the encoded size exceeds the limit, an error is returned.
+    fn encode_max(&self, size: usize) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error>;
+    // Encodes the data, truncating the encoded message if it exceeds the specified size limit.
+    /// The resulting data will fit within the limit, but may be incomplete if truncation occurs.
+    fn encode_truncate(&self, limit: usize) -> Vec<u8>;
+}
```
Comment (Rjected) [inline crates/net/eth-wire/src/types/message.rs:226]: This trait could be very useful, but the docs should probably describe more how `encode_truncate` should be used for "`vec`-like" things, because naively truncating an encoding could produce an incorrect encoding

**Review by Rjected — CHANGES_REQUESTED** [review]
Supportive of this, would like to see at least some documentation changes, to make it very clear to implementors how this should be used

File: docs/crates/network.md
Comment (mattsse) [inline docs/crates/network.md:1]: honestly we should just get rid of the code examples, they aren't really useful.

File: crates/net/network/src/eth_requests.rs [hunk 7]
```diff
@@ -216,10 +207,14 @@ where
...
-                if total_bytes > SOFT_RESPONSE_LIMIT {
-                    break
+                match receipts.encode_max(SOFT_RESPONSE_LIMIT) {
+                    Ok(_) => {
+                        // If encode_max succeeds, continue accumulating receipts
+                    }
+                    Err(_) => {
+                        // If encode_max fails,stop the loop
+                        break;
+                    }
```
Comment (mattsse) [inline crates/net/network/src/eth_requests.rs:217]: if we want to keep approximations then we don't need this here, otherwise this would just be more additional work.

File: crates/net/eth-wire/src/types/message.rs [hunk 8]
```diff
@@ -214,6 +213,78 @@ impl EthMessage {
...
+            let current_len = buffer.len();
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                buffer.truncate(current_len);
+                break;
+            }
+        }
+        buffer
+    }
+}
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:286]: this isn't quite correct because this lacks the list header. see encodeable impl for Vec

**Review by mattsse — CHANGES_REQUESTED** [review]
@emhane I'm not quite sure if we should stick with the approximation based approach or just do encoding while we're handling the response to get the accurate length

wdyt

**Comment by supernovahs** [comment]
waiting for emhane's response, then fix the nits

File: docs/crates/network.md
Comment (emhane) [reply docs/crates/network.md:1]: executable examples are good, since ci tells you what is outdated then. also why i like to link all type names in docs and avoid ever referring literally to variable names in docs, because somewhere down the line they change and docs aren't updated, which is then **so** confusing.

File: crates/net/eth-wire/src/types/message.rs (hunk 8 above)
Comment (emhane) [reply crates/net/eth-wire/src/types/message.rs:310]: this is less expensive if you allocate some capacity to the vector first, and then call `shrink_to_fit` on it. rust collections grow in powers of 2, so the first elements will be costly to reallocate. let's use the length approximation to call `Vec::with_capacity`. here it makes sense to use `smallvec` though I think, will go much faster if it the buffer stays on the stack.

**Comment by emhane** [comment]
> @emhane I'm not quite sure if we should stick with the approximation based approach or just do encoding while we're handling the response to get the accurate length
> 
> wdyt

let's keep the encoding trait and get the actual length. let's use the approximation to allocate capacity for the buffer.

**Comment by mattsse** [comment]
friendly bump here @supernovahs @emhane

**Comment by supernovahs** [comment]
Let me know if this makes sense

File: crates/net/eth-wire/src/types/message.rs [hunk 9]
```diff
@@ -286,7 +286,11 @@ pub trait EncodableExt {
 impl<T: Encodable> EncodableExt for Vec<T> {
     fn encode_max(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error> {
         let mut buffer = Vec::new();
+
         for item in self {
+            let mut header = Header { list: true, payload_length: 0 };
+            header.payload_length += item.length();
+            header.encode(&mut buffer);
```
Comment (Rjected) [inline crates/net/eth-wire/src/types/message.rs]: the item will probably also have a header, a header doesn't need to be added before each element

**Review by Rjected — CHANGES_REQUESTED** [review]
we definitely need one or two example tests for what we're expected to see. I can help out with that if you'd like

File: crates/net/eth-wire/src/types/message.rs (hunk 9 above)
Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs]: I have made changes to above code. Can u check . then proceed for tests

Comment (emhane) [reply crates/net/eth-wire/src/types/message.rs]: it looks great!

File: crates/net/eth-wire/src/types/message.rs [hunk 10]
```diff
@@ -237,6 +237,99 @@ impl EthMessage {
...
+        header.encode(&mut buffer);
+        for item in self {
+            let current_len = buffer.len();
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                buffer.truncate(current_len);
+                break;
+            }
+        }
+        buffer
```
Comment (emhane) [inline crates/net/eth-wire/src/types/message.rs:329]: please `shrink_to_fit` the buffer here. otherwise the allocated memory for the buffer will stay the approx size, even if it's filled with viewer elements! that would result in a logic bug if you are packing a response based on the length (number of elements) in the vector.

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/net/eth-wire/src/types/message.rs [hunk 11]
```diff
@@ -237,6 +237,100 @@ impl EthMessage {
...
+        }
+        header.encode(&mut buffer);
+
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
+            }
+        }
+        Ok(buffer)
```
Comment (emhane) [inline crates/net/eth-wire/src/types/message.rs:311]: `shrink_to_fit` is relevant here too, don't you agree?

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/net/eth-wire/src/types/message.rs (hunk 11 above)
Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs:311]: Added. Was thinking if its needed or not in both , since `encode_max ` will be used by user with approx > limit in most cases I think

Comment (emhane) [reply crates/net/eth-wire/src/types/message.rs:311]: better safe than sorry ;)

File: crates/net/eth-wire/src/types/message.rs [hunk 12]
```diff
@@ -301,13 +301,13 @@ impl<T: Encodable> EncodableExt for Vec<T> {
             header.payload_length += item.length();
         }
         header.encode(&mut buffer);
-
         for item in self {
             item.encode(&mut buffer);
             if buffer.len() > limit {
                 return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
```
Comment (emhane) [inline crates/net/eth-wire/src/types/message.rs:307]: asides this, the two methods are identical (`let current_len = buffer.len();` in the other trait method can be moved down to the line after the conditional statement). can we instead have a third method `encode_until_limit` with this code which returns `(Vec<u8>, bool)`. then `encode_max` will be smgth like
```rust
 fn encode_max(
        &self,
        approx: usize,
        limit: usize,
    ) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error> {
    let (mut buf, limit_exceeded) = self.encode_until_limit(aprox, limit);
    
    if limit_exceeded {
        return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
    }
    
    buf.shrink_to_fit();
    
    Ok(buf)
 }

```
and encode_truncate can be simplified too then. that way we don't have duplicate code, and it's easier to see how the methods' behaviour is different.

**Review by emhane — CHANGES_REQUESTED** [review]
lgtm. some duplicate code to take care of still.

File: crates/net/eth-wire/src/types/message.rs (hunk 5 above)
Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs]: @emhane  regarding `let current_len = buffer.len();` after the conditional statement you said. Above maybe relevant

Comment (emhane) [reply crates/net/eth-wire/src/types/message.rs]: yes I meant after the line of code, inside the scope opened by the `true` branch of the conditional statement

**Comment by supernovahs** [comment]
Let me know if this looks good

File: crates/net/eth-wire/src/types/message.rs [hunk 13]
```diff
@@ -287,6 +289,26 @@ pub trait EncodableExt {
     /// Users of this method should be aware that the truncated data may not represent the complete
     /// original dataset.
     fn encode_truncate(&self, approx: usize, limit: usize) -> Vec<u8>;
+
+    /// Encodes the data up to the specified size limit, returning the encoded data along with
+    /// a flag indicating whether the limit was exceeded. This method is useful for partial
+    /// encoding scenarios where it is necessary to encode as much as possible without going
+    /// over the limit.
```
Comment (emhane) [inline crates/net/eth-wire/src/types/message.rs:296]:
```suggestion
    /// Encodes the data up to the specified size limit, returning the encoded data along with
    /// a flag indicating whether the limit was exceeded.
```

think the use case is obvious from the first sentence :)

File: crates/net/eth-wire/src/types/message.rs [hunk 14]
```diff
@@ -287,6 +289,26 @@ pub trait EncodableExt {
...
+    ///   used to pre-allocate memory for efficiency.
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * The encoded data as `Vec<u8>`.
+    /// * The length of the data at the point of reaching the limit, or the 0 length if the limit
+    ///   was not exceeded.
+    /// * A boolean flag that is `true` if the size limit was exceeded and `false` otherwise.
```
Comment (emhane) [inline crates/net/eth-wire/src/types/message.rs:308]: this is not needed. the caller's context can just check if the length of the data before limit was exceeded is zero

File: crates/net/eth-wire/src/types/message.rs [hunk 15]
```diff
@@ -287,6 +289,26 @@ pub trait EncodableExt {
...
+    /// * `approx` - The approximate size to allocate in memory for the encoding process. This is
+    ///   used to pre-allocate memory for efficiency.
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * The encoded data as `Vec<u8>`.
+    /// * The length of the data at the point of reaching the limit, or the 0 length if the limit
+    ///   was not exceeded.
```
Comment (emhane) [inline crates/net/eth-wire/src/types/message.rs:307]:
```suggestion
    /// * The length of the data at the point just before reaching the limit, or 0 if the limit
    ///   was not exceeded.
```

File: crates/net/eth-wire/src/types/message.rs [hunk 16]
```diff
@@ -322,12 +338,24 @@ impl<T: Encodable> EncodableExt for Vec<T> {
...
+        // Return current_len = 0 as placeholder
+        (buffer, 0, false)
+    }
+
+    fn encode_truncate(&self, approx: usize, limit: usize) -> Vec<u8> {
+        let (mut buf, _current_len, limit_exceeded) = self.encode_until_limit(approx, limit);
+
+        if limit_exceeded {
+            buf.truncate(_current_len);
+            buf.shrink_to_fit(); // shrink buffer to catch cases where approx > limit
```
Comment (emhane) [inline crates/net/eth-wire/src/types/message.rs:351]: nice with the comment

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/net/eth-wire/src/types/message.rs (hunk 14 above)
Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs:308]: Not sure what you mean here :)

Comment (emhane) [reply crates/net/eth-wire/src/types/message.rs:308]: left this suggestion some lines lower but now that a commit was pushed it jumped up.


the boolean isn't needed.

**Review by emhane — CHANGES_REQUESTED** [review]
very nice! a test for `encode_truncate` and then we are good to go. please also double check that `docs/crates/network.md` is coherent with the latest changes you made.

File: crates/net/eth-wire/src/types/message.rs [hunk 17]
```diff
@@ -237,6 +237,124 @@ impl EthMessage {
...
+            header.payload_length += item.length();
+        }
+        header.encode(&mut buffer);
+        for item in self {
+            let current_len = buffer.len();
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return (buffer, current_len);
+            }
+        }
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:340]: this includes an invalid header if items are skipped
we either need to write the header last or replace in case buffer.len() > limit

**Review by mattsse — CHANGES_REQUESTED** [review]
please add roundtrip tests because currently this returns malformed rlp headers

File: crates/net/eth-wire/src/types/message.rs (hunk 17 above)
Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs:340]: Got it

Comment (Rjected) [reply crates/net/eth-wire/src/types/message.rs:340]: it should be possible to just use `.length()` to determine how many elements to write and calculate the header payload len beforehand

Comment (mattsse) [reply crates/net/eth-wire/src/types/message.rs]: that's the best solution yes

Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs]: Pushed a change, it assumes the list header is <= 3

**Comment by supernovahs** [comment]
Let me know what do you think of this impl

**Comment by supernovahs** [comment]
looks like ci issue for failing tests.working on my local

File: crates/net/eth-wire/src/types/message.rs [hunk 18]
```diff
@@ -237,6 +237,159 @@ impl EthMessage {
...
+    fn encode_until_limit(&self, approx: usize, limit: usize) -> (Vec<u8>, usize, usize) {
+        let mut buffer = Vec::with_capacity(approx);
+        let mut header = Header { list: true, payload_length: 0 };
+        for item in self {
+            // initial payload length calculation if limit is not exceeded
+            header.payload_length += item.length();
+        }
+        // header.encode(&mut buffer);
+        let header_placeholder_size = 3;
+        buffer.extend(vec![0; header_placeholder_size]);
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:349]: where's this coming from?

File: crates/net/eth-wire/src/types/message.rs [hunk 19]
```diff
@@ -237,6 +237,159 @@ impl EthMessage {
...
+        header.encode(&mut header_buffer);
+        // Ensure the encoded header does not exceed the placeholder size
+        assert!(header_buffer.len() <= 3); // header placeholder size is 3
+                                           // Replace the placeholder with the real header
+        buf[..header_buffer.len()].copy_from_slice(&header_buffer[..]);
+        // If the real header is smaller than the placeholder, remove the excess
+        if header_buffer.len() < 3 {
+            let excess = 3 - header_buffer.len();
+            buf.drain(header_buffer.len()..excess);
+        }
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:377]: this looks a bit complicated

**Review by mattsse — CHANGES_REQUESTED** [review]
I think this can be simplified a lot by following @Rjected suggestion:

we already know how many encoded items will fit because we calculate the payload_length

so the first step should be:
1. compute the header and check how large the payload length gets
2. exit before the payload length exceeds the targeted size
3. append the right amount of items to the buffer

File: crates/net/eth-wire/src/types/message.rs (hunk 18 above)
Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs:349]: Assumed max header length is <=3

Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs:349]: could also assume more, like 4 or 5 .

Comment (Rjected) [reply crates/net/eth-wire/src/types/message.rs]: the header also has a length method, it does require the payload length to be calculated properly, but in any case, we shouldn't need to extend anything at this point or include any placeholders

Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs]: yup. Made a push to remove placeholder

**Review by emhane — APPROVED** [review]

File: crates/net/eth-wire/src/types/message.rs [hunk 20]
```diff
@@ -237,6 +237,143 @@ impl EthMessage {
...
+            }
+
+            current_len = buffer.len();
+        }
+        let mut head: Vec<u8> = Vec::new();
+        header.payload_length = actual_payload_length;
+        header.encode(&mut head);
+        head.append(&mut buffer);
+        // Return current_len as 0
+        (head, 0)
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:353]: this looks very complex,

I don't think this is necessary if we first check how much items we can fit into the buffer

ptal at the solution @Rjected suggested

File: crates/net/network/src/eth_requests.rs [hunk 21]
```diff
@@ -128,13 +127,18 @@ where
...
-                if total_bytes > SOFT_RESPONSE_LIMIT {
-                    break
+                match headers.encode_max(APPROX_HEADER_SIZE, SOFT_RESPONSE_LIMIT) {
+                    Ok(_) => {
+                        // If encode_max succeeds, continue accumulating headers
+                    }
+                    Err(_) => {
+                        // If encode_max fails,stop the loop
+                        break;
+                    }
```
Comment (mattsse) [inline crates/net/network/src/eth_requests.rs:141]: this entirely defeats the purpose of using approximations because now we're spending more time on redundant encoding.

please undo these changes here and to blocks

**Review by mattsse — CHANGES_REQUESTED** [review]
ptal @Rjected

File: crates/net/eth-wire/src/types/message.rs (hunk 20 above)
Comment (supernovahs) [reply crates/net/eth-wire/src/types/message.rs:353]: need a clarification 
by limit we mean the list header + the encoding right ? 
If yes, then , header length changes depending on what the payload_length is , ex: if header.payload_length = 0 , encoded list header's len will be 1 , and 3 in other cases iirc . That is why I am calculating the list header length above.
Ex: if expected payload length is 508, limit is 511 , the `if ` condition will not pass (since 508 is not > 511), but If we include the list header len(3 length as payload_length = 508), then it will pass.

And If you mean `limit` excludes the list header, then we can indeed calculate how many items will fit beforehand correctly.

Comment (Rjected) [reply crates/net/eth-wire/src/types/message.rs:353]: the algorithm is basically:
you have a limit
you have a list of encodable items
you have a "current" _list_ length and item (you could track an index)
for each item:
* new_list_length = current_list_length + item.length()
* if Header { list: true, new_list_length }.length() + new_list_length > limit, return
* otherwise proceed

Once you know which element you need to stop at, now you should encode, you encode the list of elements up to that item
